package service;
option cc_generic_services = true;

message Identifier {
	required int32 playernum = 1;
	required int32 uniqueId = 2;
}

message NullResponse {
}

message RecordRequest {
	required Identifier identifier = 1;
	required string name = 2;
	required string mapname = 3;
	required string map_checksum = 4;
	required int32 mstime = 5;
	required int32 df_gametype = 6;
	required int32 df_promode = 7;
	required int32 df_mode = 8;
	required int32 interference = 9;
	required int32 obs_enabled = 10;
	required int32 defrag_version = 11;
}

message ClientInfoRequest {
	required Identifier identifier = 1;
	required int32 userId = 2;
	required string userinfostring = 3;
}

message ClientCommandRequest {
	required Identifier identifier = 1;
	required string command = 2;
	repeated string args = 3;
}

message LoginRequest {
	required Identifier identifier = 1;
	required string q3dfkey = 2;
}

message LoginResponse {
	required Identifier identifier = 1;
	enum LoginResult {
		FAILD = 0;
		PASSED = 1;
	}
	required LoginResult result = 2;
	required string hash = 3;
	required int32 userId = 4;
}

message ClientCommandResponse {
	required Identifier identifier = 1;
	optional string messageToPrint = 2;
}

message PrintfRequest {
	required string msg = 1;
}

service Q3dfApi {
	rpc ClientConnected (ClientInfoRequest) returns (NullResponse);
	rpc ClientDisconnected (ClientInfoRequest) returns (NullResponse);
	rpc ClientCommand (ClientCommandRequest) returns (ClientCommandResponse);
	rpc Login (LoginRequest) returns (LoginResponse);
	rpc Printf(PrintfRequest) returns (NullResponse);
	rpc SaveRecord (RecordRequest) returns (NullResponse);
}
